### DISCUSSION TITLE
Strudel Workshop

### DISCUSSION THREAD

**Discussion by jarmitage** [tidalcycles/strudel#81] [Ideas]
@yaxu and myself are giving live coding workshops in mid/late May, and we are both interested in teaching Strudel.

Here are some starting questions:

- What is a Strudel workshop and how is it different to a Tidal workshop?
- Should the broader Tidal ecosystem be introduced, and if so how, or is that overwhelming?
- Are there negative effects of only being exposed to Strudel and not Haskell Tidal?
- Is it possible / desirable to port existing Tidal workshop content to Strudel?
- Where should such pedagogical Strudel content exist such that all potential teachers & students can easily find it & collectively improve on it & translate it?
- How can/should workshop materials be delivered in Strudel? (Copy/paste? Interactive tutorial?)
- What unique affordances does Strudel have that could be pedagogically exploited? (Real-time collaboration? Easy code sharing via URLs?)

**Comment by yaxu** [comment]
Also, how to teach something that is changing fast?

**Comment by yaxu** [comment]
Another thought, especially at this stage, workshops are a great opportunity to see how people use strudel. We could make a list of things to note:
* What do they try that doesn't work, but could?
* What questions do they ask that don't yet have a good answer?
* How do they feel about live coding before, vs after the workshop? (we could make a short survey for this)

**Comment by yaxu** [comment]
* What is a Strudel workshop and how is it different to a Tidal workshop?

I think the main thing is that there is less installation stress.

* Should the broader Tidal ecosystem be introduced, and if so how, or is that overwhelming?

I think it's good to have a bit of blah blah with signposting to different systems, and that should include Tidal. Maybe there's a danger in making people think this isn't the 'real' tidal, or a cut down version of it.. I guess let them know that they're the first wave of people being introduced to strudel and that it isn't really defined what it is/how it should be used so we're super interested in their honest feedback (good and bad).

I always try to cram too much in to a workshop.

* Are there negative effects of only being exposed to Strudel and not Haskell Tidal?

There are tradeoffs, but I don't see negative effects.

The biggest downside is that strudel is changing fast, so what they learn might go out of date quickly.

* Is it possible / desirable to port existing Tidal workshop content to Strudel?

Yes not a bad idea.

* Where should such pedagogical Strudel content exist such that all potential teachers & students can easily find it & collectively improve on it & translate it?

Maybe a strudel-workshop github repo?

* How can/should workshop materials be delivered in Strudel? (Copy/paste? Interactive tutorial?)

It could be a list of URLs to different examples to go through together? and/or going through the online tutorial? Although that's trailing a bit behind at this point.

* What unique affordances does Strudel have that could be pedagogically exploited? (Real-time collaboration? Easy code sharing via URLs?)

The built-in highlighting is a big one.

**Comment by jarmitage** [comment]
All good points!

While URLs are easy to keep and share (related: https://github.com/tidalcycles/strudel/discussions/55), I can imagine that loading the entire page every time might get tiring if the pedagogical examples are quite atomic, as they often are. Are there currently plans to introduce the option of block-based evaluation? That would make a single workshop URL possible.

If that's not in the works, then it looks like updating https://strudel.tidalcycles.org/tutorial and breaking it out into separate pages (chapters?) might be the way to go for now.

Regarding https://github.com/tidalcycles/strudel/discussions/92, it would be great if documentation generated by source comments could include examples, and produce a page similar to https://strudel.tidalcycles.org/tutorial. In fact, it would be even more brilliant if the source comments and examples could appear in the REPL as a sidebar or via tooltips! This would mean that workshop content could focus on broader / higher level topics.

If we can make up our minds about which way to go in the short term, I'm sure we could draft a mini curriculum and make some good content.

**Comment by felixroos** [comment]
> it would be even more brilliant if the source comments and examples could appear in the REPL as a sidebar or via tooltips

this! we can probably get more insight from [uncle bret](http://worrydream.com/#!/LearnableProgramming). I also imagine it would be great to use the syntax tree of the user code to read the actual params and use them for a more dynamic explanation in the sidebar/tooltip. 

Another really helpful thing would be to be able reason backwards from the event output. So for example, you select an event in the pianoroll (or whatever), and the tooltip/sidebar shows you the chain of transformations from start to finish. example:

```js
"0 4".off(1/4, add(2)).scale('A minor')
```

yields events with values 

- `a3` = `"0".scale('A minor')`
- `c4` = `"0".add(2).scale('A minor')` = `"2".scale('A minor')`
- `e4` = `"4".scale('A minor')`
- `g4` = `"4".add(2).scale('A minor')` = `"6".scale('A minor')`

similar transformation chains could be generated for time structure

> Are there currently plans to introduce the option of block-based evaluation?

see https://github.com/tidalcycles/strudel/issues/34

> That would make a single workshop URL possible.

Another way to go would be to write a custom MDX file that is tailored for the workshop, where you can drop mini repls, (and possibly other components that aid explanation). The MDX could be built similar to the tutorial and made available under some subpath. I think this is even cooler than using block based evaluation, because you can drop in explanations in the form of text / image / video / visualizations / whatever between the mini repls.

**Reply by felixroos** [reply]
https://codemirror.net/6/examples/tooltip/ + https://codemirror.net/6/examples/autocompletion/

**Reply by felixroos** [reply]
https://github.com/tidalcycles/strudel/discussions/92#discussioncomment-2676602
